ext/endpoint_guessing.c

```diff
@@ -0,0 +1,237 @@
+#include "endpoint_guessing.h"
+
+#include <Zend/zend_smart_str.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zend_hash.h>
+#include <zend_string.h>
+#include <zend_types.h>
+#include "configuration.h"
+#include "ddtrace.h"
+#include "span.h"
+#include <php.h>
+#include <ext/standard/url.h>
+
+#define MAX_COMPONENTS 8
+
+static inline bool is_digit(char c) { return c >= '0' && c <= '9'; }
+
+static inline bool is_nonzero_digit(char c) { return c >= '1' && c <= '9'; }
+
+static inline bool is_hex_alpha(char c) { return (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F'); }
+
+static inline bool is_delim(char c) { return c == '.' || c == '_' || c == '-'; }
+
+static inline bool is_str_special(char c) {
+    return c == '%' || c == '&' || c == '\'' || c == '(' || c == ')' || c == '*' || c == '+' || c == ',' || c == ':' || c == '=' || c == '@';
+}
+
+/*
+{param:int}     [1-9][0-9]+                   len≥2, digits only, first 1–9
+{param:int_id}  (?=.*[0-9])[0-9._-]{3,}       len≥3, [0-9._-], must contain digit
+{param:hex}     (?=.*[0-9])[A-Fa-f0-9]{6,}    len≥6, hex digits, must contain decimal digit
+{param:hex_id}  (?=.*[0-9])[A-Fa-f0-9._-]{6,} len≥6, hex+._-, must contain decimal digit
+{param:str}     .{20,}|.*[%&'()*+,:=@].*      any chars, valid if len≥20 or contains special
+*/
+
+typedef enum {
+    COMPONENT_NONE = 0,
+    COMPONENT_IS_INT = 1 << 0,
+    COMPONENT_IS_INT_ID = 1 << 1,
+    COMPONENT_IS_HEX = 1 << 2,
+    COMPONENT_IS_HEX_ID = 1 << 3,
+    COMPONENT_IS_STR = 1 << 4,
+} component_type_t;
+
+static const char* component_type_to_string(component_type_t type) {
+    switch (type) {
+        case COMPONENT_IS_INT:
+            return "{param:int}";
+        case COMPONENT_IS_INT_ID:
+            return "{param:int_id}";
+        case COMPONENT_IS_HEX:
+            return "{param:hex}";
+        case COMPONENT_IS_HEX_ID:
+            return "{param:hex_id}";
+        case COMPONENT_IS_STR:
+            return "{param:str}";
+        default:
+            return "";
+    }
+}
+
+static inline uint8_t bool_to_mask(bool x) {
+    return (uint8_t)(-(int)x);  // 0 -> 0x00, 1 -> 0xFF
+}
+
+static component_type_t component_replacement(const char* path, size_t len) {
+    uint8_t viable_components = 0x1F;  // (COMPONENT_IS_STR << 1) - 1
+    bool found_special_char = false;
+    bool found_digit = false;
+
+    if (len < 2) {
```

Comment:
```suggestion
    if (len < 2) {
        ZEND_ASSERT(len > 0);
```
To make sure, rather than checking later. At least it's never to be called with len == 0.